### PR TITLE
CSMS API

### PR DESCRIPTION
**Pull request by scvannost**
## What & Why

[CIDC-1115](https://dfcijira.dfci.harvard.edu:8443/browse/CIDC-1115): Implements an endpoint that takes a CSMS-style manifest JSON and inserts it into both the JSON and relational databases

## How

`POST`ing to `/trial_metadata/new_manifest` with a CSMS-style manifest calls two separate hooks -- one for adding to the legacy JSON blobs and one for adding to the relational tables. Both process the manifest to ensure validity against the current CIDC data and then inserts new samples (and participants if needed) with all metadata passed.

## Remarks

Do not address the situation where the manifest already exists in CIDC with different values -- this is for brand-new never-before-seen manifests.

## Checklist

Please include and complete the following checklist. You can mark an item as complete with the `- [x]` prefix:

- [x] Tests - Added unit tests for new code, regression tests for bugs and updated the integration tests if required
- [x] Formatting & Linting - `black` and `flake8` have been used to ensure styling guidelines are met
- [x] Type Annotations - All new code has been type annotated in the function signatures using type hints
- [x] Docstrings - Docstrings have been provided for functions
- [ ] Documentation - README has been updated to explain major changes & new features
- [x] Package version - Manually bumped the API package version in [setup.py](https://github.com/CIMAC-CIDC/cidc-api-gae/blob/master/setup.py#L21)

### CODE REVIEW

**Comment by jim-bo** [comment]
@scvannost maybe we can review together or you could add some high level descriptions to the PR about what the major functions you've added are doing. Since its a new features I'm having trouble finding a good starting point for the review.

File: tests/resources/test_trial_metadata.py [hunk 1]
```diff
@@ -461,3 +462,61 @@ def test_get_trial_metadata_summaries(cidc_api, clean_db, monkeypatch):
...
+            validate_json_blob(md_json)
+            validate_relational("test_trial")
+
+    # reusing one returns an error
+    manifest = [m for m in manifests if m.get("status") in [None, "qc_complete"]][0]
+    res = client.post("/trial_metadata/new_manifest", json=manifest)
+    assert res.status_code != 200
+    assert "already exists for trial" in res.json["error"]
+
+    # if not qc_complete, returns an error
```
Comment (jim-bo) [inline tests/resources/test_trial_metadata.py:518]: I would add a sub-test (or a new test) around failures in the payload. Think about maybe creating a manifest where the relational constraints are violated somehow, or values are malformed (bad cimac id).

File: tests/csms/test_auth.py [hunk 2]
```diff
@@ -43,13 +43,9 @@ def test_get_with_authorization(monkeypatch):
     get_mock = MagicMock()
     monkeypatch.setattr(auth.requests, "get", get_mock)
 
-    auth.get_with_authorization(
-        "foobar", "b", "c", headers={"beep": "boop"}, test="baz"
-    )
+    auth.get_with_authorization("/foobar", headers={"beep": "boop"}, test="baz")
     get_mock.assert_called_with(
```
[the remark this answers is not in the document]
Comment (scvannost) [reply tests/csms/test_auth.py:47]: yes we did, this is just a change in the first param to add "/"

File: tests/csms/utils.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
+from sqlalchemy.orm.session import Session
```
[the remark this answers is not in the document]
Comment (scvannost) [reply tests/csms/utils.py:1]: both together are tested via `/test/resources/trial_metadata.py::test_add_new_manifest_from_json`. see other discussion about failing independently / rollbacks

File: cidc_api/resources/trial_metadata.py [hunk 4]
```diff
@@ -135,3 +139,23 @@ def update_trial_metadata_by_trial_id(trial, trial_updates):
         )
 
     return trial
+
+
+@trial_metadata_bp.route("/new_manifest", methods=["POST"])
+@requires_auth("new_manifest", [CIDCRole.ADMIN.value])
+def add_new_manifest_from_json():
```
[the remark this answers is not in the document]
Comment (scvannost) [reply cidc_api/resources/trial_metadata.py:146]: this way also allows a future possibility where CSMS can directly POST a new manifest to CIDC. we'd only have to add the auth for a csms post

File: tests/csms/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (scvannost) [reply tests/csms/utils.py:1]: to check the integrity of both, you would call validate_relational and validate_blob with the same trial_id

File: cidc_api/resources/trial_metadata.py [hunk 5]
```diff
@@ -135,3 +139,23 @@ def update_trial_metadata_by_trial_id(trial, trial_updates):
...
+    try:
+        # relational hook
+        insert_manifest_from_json(request.json)
+
```
[the remark this answers is not in the document]
Comment (scvannost) [reply cidc_api/resources/trial_metadata.py:150]: deferred per discussion in MM

File: tests/csms/utils.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+                    "version": 1,
+                    "txTime": f"2021-01-03T00:00:00Z",
+                    "txId": "0KJLIHGFEDCBA987654321",
+                },
+            }
+        )
+
+
+@with_default_session
+def validate_relational(trial_id: str, *, session: Session):
```
[the remark this answers is not in the document]
Comment (scvannost) [reply tests/csms/utils.py:99]: in next commit

File: tests/csms/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (scvannost) [reply tests/csms/utils.py:1]: deferred per discussion in MM

File: tests/resources/test_trial_metadata.py (hunk 1 above)
Comment (scvannost) [reply tests/resources/test_trial_metadata.py:518]: so relational constraints are violated in the test cases where the cohort or collection_event name are in the DB, via `tests/models/templates/test_csms_api.py`. malformed CIMAC IDs aren't explicitly tested, but are checked [here](https://github.com/CIMAC-CIDC/cidc-api-gae/pull/576/files#diff-f419100f9b0f7250adca8e1161d94fab18150de16ddf08fdec0df52053a6344fR228)

**Review by jim-bo — APPROVED** [review]
Offline discussion with @scvannost and we decided while nice, it might not be necessary to group the RDB and JSON changes in a single transaction. If they get out of sync we can manually re-build the RDB for now.

File: tests/csms/test_auth.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (scvannost) [reply tests/csms/test_auth.py:47]: as `auth.CSMS_BASE_URL` used below doesn't have a trailing `/`

**Review by roshni-b — APPROVED** [review]

**Pull request merged**